ampd/src/mvx/verifier.rs

```diff
@@ -0,0 +1,608 @@
+        let destination_chain = String::from_utf8(destination_chain)?;
+        if destination_chain != self.destination_chain.to_string() {
+            return Ok(false);
+        }
+
+        let destination_address = topics.get(3).ok_or("")?;
+        let destination_address = STANDARD.decode(destination_address)?;
+        let destination_address = String::from_utf8(destination_address)?;
+        if destination_address != self.destination_address {
+            return Ok(false);
+        }
+
+        let payload_hash = topics.get(4).ok_or("")?;
+        let payload_hash = STANDARD.decode(payload_hash)?;
+        if payload_hash.len() != 32
+            || Hash::from_slice(payload_hash.as_slice()) != self.payload_hash
+        {
+            return Ok(false);
+        }
+
+        return Ok(true);
+    }
+}
+
+impl VerifierSetConfirmation {
+    fn eq_event(&self, event: &Events) -> Result<bool, Box<dyn std::error::Error>> {
+        if event.identifier != ROTATE_SIGNERS_IDENTIFIER {
+            return Ok(false);
+        }
+
+        let topics = event.topics.as_ref().ok_or("")?;
+
+        let event_name = topics.get(0).ok_or("")?;
+        let event_name = STANDARD.decode(event_name)?;
+        if event_name.as_slice() != SIGNERS_ROTATED_EVENT.as_bytes() {
+            return Ok(false);
+        }
+
+        let signers_hash = topics.get(2).ok_or("")?;
+        let signers_hash = STANDARD.decode(signers_hash)?;
+
+        let weighted_signers = WeightedSigners::from(&self.verifier_set);
+
+        if signers_hash.len() != 32 || signers_hash.as_slice() != weighted_signers.hash().as_slice()
+        {
+            return Ok(false);
+        }
+
+        return Ok(true);
+    }
+}
+
+fn find_event<'a>(
+    transaction: &'a TransactionOnNetwork,
+    gateway_address: &Address,
+    log_index: usize,
+    identifier: &str,
+    needed_event_name: &[u8],
+) -> Option<&'a Events> {
+    if transaction.logs.is_none() {
+        return None;
+    }
+
+    // We only support log index being 0, so one supported event per transaction because of MultiversX limitations
+    if log_index != 0 {
+        return None;
+    }
+
+    // Because of current relayer limitation, if log_index is 0, we will search through the logs and get the first log
+    // which corresponds to the gateway address, hence only supporting one cross chain call per transaction
+
+    for event in transaction.logs.as_ref().unwrap().events.iter() {
+        if event.address.to_bytes() == gateway_address.to_bytes() && event.identifier == identifier
+        {
+            let topics = unwrap_or_continue!(event.topics.as_ref());
+
+            let event_name = unwrap_or_continue!(topics.get(0));
+            let event_name = STANDARD.decode(event_name).unwrap_or(Vec::new());
+            if event_name.as_slice() == needed_event_name {
+                return Some(event);
+            }
+        }
+    }
+
+    return None;

```

Comment:
no plain `unwrap()` in production code. You can get around it by making use of the `?` operator. That also saves you the use of a macro.

```suggestion
    // We only support log index being 0, so one supported event per transaction because of MultiversX limitations
    if log_index != 0 {
        return None;
    }

    // Because of current relayer limitation, if log_index is 0, we will search through the logs and get the first log
    // which corresponds to the gateway address, hence only supporting one cross chain call per transaction
    transaction
        .logs
        .as_ref()?
        .events
        .iter()
        .filter(|events| {
            events.address.to_bytes() == gateway_address.to_bytes()
                && events.identifier == identifier
        })
        .filter_map(|events| {
            let event_name = events.topics.as_ref()?.get(0)?;
            let event_name = STANDARD.decode(event_name).ok()?;
            if event_name.as_slice() == needed_event_name {
                Some(events)
            } else {
                None
            }
        })
        .next()
```

Reply:
Done.